File: python/packages/jumpstarter-cli/jumpstarter_cli/shell.py
import logging
import sys
from contextlib import ExitStack
from datetime import timedelta

import anyio
import click
from anyio import create_task_group, get_cancelled_exc_class
from jumpstarter_cli_common.config import opt_config
from jumpstarter_cli_common.exceptions import find_exception_in_group, handle_exceptions_with_reauthentication
from jumpstarter_cli_common.oidc import (
    TOKEN_EXPIRY_WARNING_SECONDS,
    format_duration,
    get_token_remaining_seconds,
)
from jumpstarter_cli_common.signal import signal_handler

from .common import opt_acquisition_timeout, opt_duration_partial, opt_selector
from .login import relogin_client
from jumpstarter.client.client import client_from_path
from jumpstarter.common import ExporterStatus
from jumpstarter.common.exceptions import ConnectionError, ExporterOfflineError
from jumpstarter.common.utils import launch_shell
from jumpstarter.config.client import ClientConfigV1Alpha1
from jumpstarter.config.exporter import ExporterConfigV1Alpha1

logger = logging.getLogger(__name__)


def _run_shell_only(lease, config, command, path: str) -> int:
    """Run just the shell command without log streaming."""
    return launch_shell(
        path,
        lease.exporter_name,
        config.drivers.allow,
        config.drivers.unsafe,
        config.shell.use_profiles,
        command=command,
        lease=lease,
    )


def _warn_about_expired_token(lease_name: str, selector: str) -> None:
    """Warn user that lease won't be cleaned up due to expired token."""
    click.echo(click.style("\nToken expired - lease cleanup will fail.", fg="yellow", bold=True))
    click.echo(click.style(f"Lease '{lease_name}' will remain active.", fg="yellow"))
    click.echo(click.style(f"To reconnect: JMP_LEASE={lease_name} jmp shell", fg="cyan"))


async def _monitor_token_expiry(config, cancel_scope) -> None:
    """Monitor token expiry and warn user."""
    token = getattr(config, "token", None)
    if not token:
        return

    warned = False
    while not cancel_scope.cancel_called:
        try:
            remaining = get_token_remaining_seconds(token)
            if remaining is None:
                return

            if remaining <= 0:
                click.echo(click.style("\nToken expired! Exiting shell.", fg="red", bold=True))
                cancel_scope.cancel()
                return

            if remaining <= TOKEN_EXPIRY_WARNING_SECONDS and not warned:
                duration = format_duration(remaining)
                click.echo(
                    click.style(
                        f"\nToken expires in {duration}. Session will continue but cleanup may fail on exit.",
                        fg="yellow",
                        bold=True,
                    )
                )
                warned = True

            await anyio.sleep(30)
        except Exception:
            return


async def _run_shell_with_lease_async(lease, exporter_logs, config, command, cancel_scope):  # noqa: C901
    """Run shell with lease context managers and wait for afterLease hook if logs enabled.

    When exporter_logs is enabled, this function will:
    1. Connect and start log streaming with background status monitor
    2. Wait for beforeLease hook to complete (logs stream in real-time)
    3. Run the shell command
    4. After shell exits, call EndSession to trigger and wait for afterLease hook
    5. Logs stream to client during hook execution
    6. Release the lease after hook completes

    Uses non-blocking polling via StatusMonitor for robust status tracking.
    If Ctrl+C is pressed during EndSession, the wait is skipped but the lease is still released.
    """
    async with lease.serve_unix_async() as path:
        async with lease.monitor_async():
            # Use ExitStack for the client (required by client_from_path)
            with ExitStack() as stack:
                async with client_from_path(
                    path, lease.portal, stack, allow=lease.allow, unsafe=lease.unsafe
                ) as client:
                    # Probe GetStatus before log stream so the server-side error
                    # from unsupported exporters is not streamed to the terminal.
                    await client.get_status_async()

                    # Start log streaming and status monitor together
                    # The status monitor polls in the background for reliable status tracking
                    async with client.log_stream_async(show_all_logs=exporter_logs):
                        async with client.status_monitor_async(poll_interval=0.3) as monitor:
                            # Wait for beforeLease hook to complete while logs are streaming
                            # This allows hook output to be displayed in real-time
                            # Uses non-blocking polling instead of streaming for robustness
                            logger.info("Waiting for beforeLease hook to complete...")

                            # Wait for LEASE_READY or hook failure using background monitor
                            result = await monitor.wait_for_any_of(
                                [ExporterStatus.LEASE_READY, ExporterStatus.BEFORE_LEASE_HOOK_FAILED], timeout=300.0
                            )

                            if result == ExporterStatus.BEFORE_LEASE_HOOK_FAILED:
                                reason = monitor.status_message or "beforeLease hook failed"
                                raise ExporterOfflineError(reason)
                            elif result is None:
                                if monitor.connection_lost:
                                    # Connection lost while waiting for hook — lease expired
                                    logger.info("Lease expired while waiting for beforeLease hook to complete")
                                    return 0
                                else:
                                    reason = monitor.status_message or "Timeout waiting for beforeLease hook"
                                    raise ExporterOfflineError(reason)

                            logger.debug("Exporter ready (status: %s), launching shell...", result)

                            # Run the shell command
                            exit_code = await anyio.to_thread.run_sync(_run_shell_only, lease, config, command, path)

                            # Shell has exited. For auto-created leases (release=True), call
                            # EndSession to trigger afterLease hook while keeping log stream
                            # and status monitor open. For pre-created leases (release=False),
                            # skip EndSession so the exporter stays in LEASE_READY and the
                            # user can reconnect later.
                            if (
                                lease.release
                                and lease.name
                                and not cancel_scope.cancel_called
                                and not monitor._get_status_unsupported
                            ):
                                # Quick probe to catch exporter restarts the slow-poll loop
                                # (5s interval in LEASE_READY) may not have detected yet.
                                if not monitor.connection_lost:
                                    try:
                                        probe_status = await client.get_status_async()
                                        if probe_status is not None and probe_status not in (
                                            ExporterStatus.LEASE_READY,
                                            ExporterStatus.AFTER_LEASE_HOOK,
                                        ):
                                            logger.debug(
                                                "Exporter in unexpected state (%s), skipping afterLease hook",
                                                probe_status,
                                            )
                                            monitor._connection_lost = True
                                    except Exception:
                                        logger.debug("Connection probe failed, marking connection as lost")
                                        monitor._connection_lost = True

                                if monitor.connection_lost:
                                    logger.debug("Connection already lost, skipping afterLease hook")
                                else:
                                    logger.info("Running afterLease hook (Ctrl+C to skip)...")
                                    try:
                                        # EndSession triggers the afterLease hook asynchronously
                                        # Wrap in anyio timeout as safety net in case gRPC deadline
                                        # doesn't fire on a broken channel (e.g. lease timeout)
                                        success = False
                                        with anyio.move_on_after(10):
                                            success = await client.end_session_async()
                                        if success:
                                            # Wait for hook to complete using background monitor
                                            # This allows afterLease logs to be displayed in real-time
                                            result = await monitor.wait_for_any_of(
                                                [ExporterStatus.AVAILABLE, ExporterStatus.AFTER_LEASE_HOOK_FAILED],
                                                timeout=30.0,
                                            )
                                            if result == ExporterStatus.AVAILABLE:
                                                logger.info("afterLease hook completed")
                                            elif result == ExporterStatus.AFTER_LEASE_HOOK_FAILED:
                                                reason = monitor.status_message or "afterLease hook failed"
                                                raise ExporterOfflineError(reason)
                                            elif monitor.connection_lost:
                                                # If connection lost during AFTER_LEASE_HOOK, the hook
                                                # likely failed and the exporter shut down (onFailure=exit)
                                                if monitor.current_status == ExporterStatus.AFTER_LEASE_HOOK:
                                                    reason = "afterLease hook failed (connection lost)"
                                                    raise ExporterOfflineError(reason)
                                                # Connection lost but hook wasn't running. This is expected when
                                                # the lease times out — exporter handles its own cleanup.
                                                logger.info("Connection lost, skipping afterLease hook wait")
                                            elif result is None:
                                                logger.warning("Timeout waiting for afterLease hook to complete")
                                        else:
                                            logger.debug("EndSession not implemented, skipping hook wait")
                                    except ExporterOfflineError:
                                        raise
                                    except Exception as e:
                                        logger.warning("Error during afterLease hook: %s", e)

                            return exit_code


async def _shell_with_signal_handling(  # noqa: C901
    config, selector, lease_name, duration, exporter_logs, command, acquisition_timeout
):
    """Handle lease acquisition and shell execution with signal handling."""
    exit_code = 0
    cancelled_exc_class = get_cancelled_exc_class()
    lease_used = None

    # Check token before starting
    token = getattr(config, "token", None)
    if token:
        remaining = get_token_remaining_seconds(token)
        if remaining is not None and remaining <= 0:
            raise ConnectionError("token is expired")

    async with create_task_group() as tg:
        tg.start_soon(signal_handler, tg.cancel_scope)

        try:
            try:
                async with anyio.from_thread.BlockingPortal() as portal:
                    async with config.lease_async(selector, lease_name, duration, portal, acquisition_timeout) as lease:
                        lease_used = lease

                        # Start token monitoring only once we're in the shell
                        tg.start_soon(_monitor_token_expiry, config, tg.cancel_scope)

                        exit_code = await _run_shell_with_lease_async(
                            lease, exporter_logs, config, command, tg.cancel_scope
                        )
            except BaseExceptionGroup as eg:
                for exc in eg.exceptions:
                    if isinstance(exc, TimeoutError):
                        raise exc from None
                offline_exc = find_exception_in_group(eg, ExporterOfflineError)
                if offline_exc:
                    raise offline_exc from None
                if lease_used is not None:
                    raise ExporterOfflineError("Connection to exporter lost") from None
                raise
            except cancelled_exc_class:
                # Check if cancellation was due to token expiry
                token = getattr(config, "token", None)
                if lease_used and token:
                    remaining = get_token_remaining_seconds(token)
                    if remaining is not None and remaining <= 0:
                        _warn_about_expired_token(lease_used.name, selector)
                exit_code = 2
        finally:
            if not tg.cancel_scope.cancel_called:
                tg.cancel_scope.cancel()

    return exit_code


@click.command("shell")
@opt_config()
@click.argument("command", nargs=-1)
# client specific
# TODO: warn if these are specified with exporter config
@click.option("--lease", "lease_name")
@opt_selector
@opt_duration_partial(default=timedelta(minutes=30), show_default="00:30:00")
@click.option("--exporter-logs", is_flag=True, help="Enable exporter log streaming")
@opt_acquisition_timeout()
# end client specific
@handle_exceptions_with_reauthentication(relogin_client)
def shell(config, command: tuple[str, ...], lease_name, selector, duration, exporter_logs, acquisition_timeout):
    """
    Spawns a shell (or custom command) connecting to a local or remote exporter

    COMMAND is the custom command to run instead of shell.

    Example:

    .. code-block:: bash

        $ jmp shell --exporter foo -- python bar.py
    """

    match config:
        case ClientConfigV1Alpha1():
            exit_code = anyio.run(
                _shell_with_signal_handling,
                config,
                selector,
                lease_name,
                duration,
                exporter_logs,
                command,
                acquisition_timeout,
            )
            sys.exit(exit_code)

        case ExporterConfigV1Alpha1():
            with config.serve_unix() as path:
                # SAFETY: the exporter config is local thus considered trusted
                launch_shell(
                    path,
                    "local",
                    allow=[],
                    unsafe=True,
                    use_profiles=False,
                    command=command,
                )

File: rust/jumpstarter-lease/src/shell.rs
//! `jmp shell` orchestration: acquire a lease, serve the transport host, run a
//! shell/command wired to it, and release the lease on exit
//! (`common/utils.py:launch_shell`, `jumpstarter-cli/.../shell.py`).
//!
//! The child inherits stdio (so an interactive shell works) and the established
//! environment contract: `JUMPSTARTER_HOST`, `JMP_DRIVERS_ALLOW`
//! (`"UNSAFE"` or the comma-joined allow-list), and `_JMP_SUPPRESS_DRIVER_WARNINGS`.

use std::time::{Duration, Instant};

use jumpstarter_config::ClientConfig;
use jumpstarter_protocol::v1::exporter_service_client::ExporterServiceClient;
use jumpstarter_protocol::v1::{EndSessionRequest, ExporterStatus, GetStatusRequest};
use tracing::{debug, info, trace, warn};

use crate::error::{ClientError, LeaseError};
use crate::lease::{acquire, CreateLeaseParams, LeaseProvider, LeaseTiming};
use crate::service::ControllerClient;
use crate::transport;

/// Prefix the exporter puts on a `GetStatus` message when a hook failed with
/// `on_failure: warn` (`common/__init__.py:HOOK_WARNING_PREFIX`). The client strips
/// it and prints `Warning: …` (`shell.py:323`).
const HOOK_WARNING_PREFIX: &str = "[HOOK_WARNING] ";

/// If `message` is a hook warning, print it as `Warning: …` (mirrors `shell.py`'s
/// yellow warning). Returns whether it printed, so callers print each warning once.
fn print_hook_warning(message: &Option<String>) -> bool {
    if let Some(text) = message
        .as_deref()
        .and_then(|m| m.strip_prefix(HOOK_WARNING_PREFIX))
    {
        eprintln!("Warning: {text}");
        true
    } else {
        false
    }
}

/// Options for [`run`].
#[derive(Debug, Clone)]
pub struct ShellOptions {
    /// Label selector for the lease.
    pub selector: Option<String>,
    /// Specific exporter name (instead of a selector).
    pub exporter_name: Option<String>,
    /// Reuse an existing lease by name (`--lease`/`JMP_LEASE`). When set, the lease
    /// is *not* released on exit so it can be reconnected to.
    pub lease_name: Option<String>,
    /// Lease duration.
    pub duration: Duration,
    /// Override the lease-acquisition timeout (`--acquisition-timeout`).
    pub acquisition_timeout: Option<Duration>,
    /// Stream *all* exporter logs (driver + system) during the session. Hook logs
    /// (before/afterLease) are always streamed regardless (`--exporter-logs`).
    pub exporter_logs: bool,
    /// Command to run instead of an interactive `$SHELL`.
    pub command: Option<Vec<String>>,
}

impl Default for ShellOptions {
    fn default() -> Self {
        Self {
            selector: None,
            exporter_name: None,
            lease_name: None,
            duration: Duration::from_secs(30 * 60),
            acquisition_timeout: None,
            exporter_logs: false,
            command: None,
        }
    }
}

/// Acquire a lease for `config`, serve the transport socket, run the shell/command,
/// release the lease (only if we created it), and return the child's exit code.
pub async fn run(config: &ClientConfig, opts: ShellOptions) -> Result<i32, ClientError> {
    let client = ControllerClient::connect(config).await?;

    // A reused lease (`--lease`/resolved-from-active) is not released on exit
    // (`shell.py`: `lease.release` is False for pre-created leases).
    let release = opts.lease_name.is_none();
    let timing = LeaseTiming {
        acquisition_timeout: opts
            .acquisition_timeout
            .unwrap_or_else(|| LeaseTiming::default().acquisition_timeout),
        ..LeaseTiming::default()
    };

    let acquired = match acquire(
        &client,
        CreateLeaseParams {
            selector: opts.selector.clone(),
            exporter_name: opts.exporter_name.clone(),
            duration: opts.duration,
            ..Default::default()
        },
        opts.lease_name.clone(),
        Some(&config.metadata.name),
        timing,
    )
    .await
    {
        Ok(acquired) => acquired,
        // A lease that is `Released` before it is ever observed `Ready` means the
        // exporter took the lease and gave it back during setup — most commonly a
        // `beforeLease` hook failing with `onFailure: endLease`, which ends the
        // lease faster than the acquisition poll catches the brief `Ready=True`
        // window (a fast controller widens this race; the exporter/lease genuinely
        // went away). Surface it like the beforeLease-wait's own connection-loss
        // path (see `wait_for_lease_ready`) rather than the raw "lease released",
        // so the user sees why the shell could not start.
        Err(ClientError::Lease(LeaseError::Released(name))) => {
            warn!(
                lease = %name,
                "lease released before it became ready; treating as connection loss"
            );
            return Err(ClientError::Exporter(
                "Connection to exporter lost".to_string(),
            ));
        }
        Err(e) => return Err(e),
    };

    // The exporter name labels the shell prompt + the lifecycle messages (`shell.py`
    // passes `lease.exporter_name` to `launch_shell`); fall back to the lease name.
    let context = if acquired.exporter.is_empty() {
        acquired.name.clone()
    } else {
        acquired.exporter.clone()
    };
    info!(lease = %acquired.name, exporter = %context, "acquired lease");
    eprintln!("Acquired lease {} on exporter {context}", acquired.name);

    // Exporter context for the shell (#53): fetch the exporter's labels so the child can read
    // `JMP_EXPORTER`/`JMP_LEASE`/`JMP_EXPORTER_LABELS` via `env_with_metadata()`. A fetch failure is
    // non-fatal (empty labels), mirroring `lease.py::_fetch_exporter_labels`.
    let exporter_labels = if acquired.exporter.is_empty() {
        String::new()
    } else {
        debug!(exporter = %acquired.exporter, "fetching exporter labels");
        match client.get_exporter(&acquired.exporter).await {
            Ok(e) => {
                let mut pairs: Vec<(String, String)> = e.labels.into_iter().collect();
                pairs.sort();
                let labels = pairs
                    .into_iter()
                    .map(|(k, v)| format!("{k}={v}"))
                    .collect::<Vec<_>>()
                    .join(",");
                debug!(exporter = %acquired.exporter, labels = %labels, "fetched exporter labels");
                labels
            }
            Err(e) => {
                warn!(exporter = %acquired.exporter, error = %e, "could not fetch exporter labels");
                String::new()
            }
        }
    };
    let lease_env = LeaseEnv {
        exporter: acquired.exporter.clone(),
        lease: acquired.name.clone(),
        labels: exporter_labels,
    };

    let host =
        transport::serve_default(client.clone(), acquired.name.clone(), config.tls.clone()).await?;
    let socket = host.jumpstarter_host();

    let drivers_allow = if config.drivers.r#unsafe {
        "UNSAFE".to_string()
    } else {
        config.drivers.allow.join(",")
    };
    let insecure = crate::channel::is_insecure(&config.tls);

    // Stream exporter logs for the whole session (`shell.py:log_stream_async`): hook
    // (before/afterLease) output is always shown; driver/system only with
    // `--exporter-logs`. Connect via the local transport proxy socket and give the
    // beforeLease replay a moment to flush before the command's own output.
    let log_task = crate::exporter_logs::spawn_controller(socket.clone(), opts.exporter_logs);

    // Wait for the beforeLease hook to finish (LEASE_READY) before running the
    // command, so a beforeLease failure surfaces as a clear error instead of the
    // child's opaque connection timeout (`shell.py`: wait for LEASE_READY /
    // BEFORE_LEASE_HOOK_FAILED via the status monitor, before running the command).
    eprintln!("Waiting for beforeLease hook to complete...");
    if let Err(e) = wait_for_lease_ready(&socket).await {
        warn!(lease = %acquired.name, error = %e, "beforeLease wait failed; tearing down session");
        log_task.abort();
        if release {
            debug!(lease = %acquired.name, "releasing lease after beforeLease failure");
            eprintln!("Releasing lease {}", acquired.name);
            if let Err(err) = client.delete_lease(&acquired.name).await {
                log_release_outcome(&acquired.name, &err);
            }
        }
        drop(host);
        return Err(e);
    }
    debug!(lease = %acquired.name, "beforeLease hook complete (lease ready)");

    let result = spawn_child(
        &opts.command,
        &socket,
        &context,
        &drivers_allow,
        insecure,
        None,
        Some(&lease_env),
    )
    .await;

    // End the session so the afterLease hook runs while the log stream is still open,
    // and wait for it to finish before releasing the lease (`shell.py`: EndSession +
    // wait-for-afterLease, then release). Only for leases we created; a reused lease
    // (`--lease`) is left intact for reconnection.
    if release {
        debug!(lease = %acquired.name, "ending session and running afterLease hook");
        eprintln!("Running afterLease hook (Ctrl+C to skip)...");
        end_session_and_wait(&socket).await;
        debug!(lease = %acquired.name, "afterLease hook completed");
        eprintln!("afterLease hook completed");
    }
    // Brief flush window for trailing afterLease lines, then stop streaming.
    tokio::time::sleep(Duration::from_millis(200)).await;
    log_task.abort();

    if release {
        debug!(lease = %acquired.name, "releasing lease");
        eprintln!("Releasing lease {}", acquired.name);
        if let Err(err) = client.delete_lease(&acquired.name).await {
            log_release_outcome(&acquired.name, &err);
        }
    }
    drop(host);

    result
}

/// Wait for the lease to become ready (`LEASE_READY`) over a *single, persistent*
/// transport-proxy connection, returning an error if the `beforeLease` hook fails.
///
/// Establishing the connection performs the controller `Dial`, which the controller
/// retries server-side only while the exporter is `Available`
/// (`controller_service.go:Dial`); a successful connect therefore means the exporter
/// already went active (`BeforeLeaseHook`+). After that, `GetStatus` on the same
/// connection drives the decision:
/// - `LEASE_READY` → `Ok` (run the command);
/// - `BeforeLeaseHookFailed` (current or `previous_status`) → the exporter's message;
/// - back to `Available`/`Offline`, or the connection drops → the lease ended without
///   becoming ready (endLease, or `onFailure: exit` shut the exporter down).
///
/// Reusing one connection is essential: reconnecting per poll would re-trigger the
/// controller's 30 s `Available` Dial-retry once the lease ends and hang the shell.
async fn wait_for_lease_ready(socket: &str) -> Result<(), ClientError> {
    const RPC_TIMEOUT: Duration = Duration::from_secs(5);
    // Bounds the initial connect (covers the controller's ~30 s Available retry).
    const CONNECT_TIMEOUT: Duration = Duration::from_secs(40);
    let connect = tokio::time::timeout(
        CONNECT_TIMEOUT,
        crate::exporter_logs::uds_channel(socket.to_string()),
    );
    let channel = match connect.await {
        Ok(Ok(ch)) => ch,
        // Never became reachable — the exporter went offline / shut down (e.g. a
        // beforeLease `onFailure: exit`) before serving.
        _ => {
            return Err(ClientError::Exporter(
                "Connection to exporter lost".to_string(),
            ))
        }
    };
    // The connect succeeded, so the exporter reached an active (BeforeLeaseHook+)
    // state; a later drop/return-to-Available means the lease ended.
    let mut client = ExporterServiceClient::new(channel);

    // beforeLease hooks can be slow (their own timeout defaults higher than 120s); give the wait
    // a generous budget so a legitimately slow hook isn't cut off.
    let deadline = Instant::now() + Duration::from_secs(300);
    while Instant::now() < deadline {
        let resp =
            match tokio::time::timeout(RPC_TIMEOUT, client.get_status(GetStatusRequest {})).await {
                Ok(Ok(resp)) => resp.into_inner(),
                // Connection dropped / RPC errored — the session ended before becoming ready.
                _ => {
                    return Err(ClientError::Exporter(
                        "Connection to exporter lost".to_string(),
                    ))
                }
            };
        let status = ExporterStatus::try_from(resp.status).unwrap_or(ExporterStatus::Unspecified);
        let prev = resp
            .previous_status
            .and_then(|p| ExporterStatus::try_from(p).ok());

        if status == ExporterStatus::LeaseReady {
            // beforeLease failed with `on_failure: warn` — show the warning, proceed.
            print_hook_warning(&resp.message);
            return Ok(());
        }
        if status == ExporterStatus::BeforeLeaseHookFailed
            || prev == Some(ExporterStatus::BeforeLeaseHookFailed)
        {
            let msg = (status == ExporterStatus::BeforeLeaseHookFailed)
                .then(|| resp.message.clone())
                .flatten()
                .filter(|m| !m.is_empty())
                .unwrap_or_else(|| "beforeLease hook failed".to_string());
            return Err(ClientError::Exporter(msg));
        }
        // The lease went active then returned to Available/Offline: it ended without
        // ever becoming ready.
        if matches!(status, ExporterStatus::Available | ExporterStatus::Offline) {
            return Err(ClientError::Exporter(
                "Connection to exporter lost".to_string(),
            ));
        }
        // BeforeLeaseHook / transitional — keep polling on the same connection.
        tokio::time::sleep(Duration::from_millis(150)).await;
    }
    // The budget elapsed without ever reaching LEASE_READY: do NOT silently proceed to run the
    // command as if the lease were ready — surface a clear timeout.
    warn!("timed out waiting for beforeLease hook to complete; the lease never became ready");
    Err(ClientError::Exporter(
        "Timed out waiting for beforeLease hook to complete".to_string(),
    ))
}

/// Ask the exporter to end the session (running the `afterLease` hook) and poll
/// `GetStatus` until it completes — so afterLease output streams to the still-open
/// log task before the lease is released (`shell.py:_run_shell_with_lease_async`).
///
/// Best-effort: every RPC is bounded by a timeout (a half-open tunnel must not hang
/// the shell), and the whole wait is capped. On any failure we return and let the
/// normal lease release proceed.
async fn end_session_and_wait(socket: &str) {
    const RPC_TIMEOUT: Duration = Duration::from_secs(5);

    let connect = tokio::time::timeout(
        RPC_TIMEOUT,
        crate::exporter_logs::uds_channel(socket.to_string()),
    );
    let channel = match connect.await {
        Ok(Ok(ch)) => ch,
        Ok(Err(e)) => {
            debug!(error = %e, "end_session: could not connect to exporter; skipping afterLease wait");
            return;
        }
        Err(_elapsed) => {
            debug!("end_session: connect to exporter timed out; skipping afterLease wait");
            return;
        }
    };
    let mut client = ExporterServiceClient::new(channel);

    match tokio::time::timeout(RPC_TIMEOUT, client.end_session(EndSessionRequest {})).await {
        Ok(Ok(_)) => debug!("EndSession requested; waiting for afterLease to complete"),
        // EndSession unsupported/unreachable — nothing to wait for.
        Ok(Err(status)) => {
            debug!(code = ?status.code(), "EndSession RPC errored; skipping afterLease wait");
            return;
        }
        Err(_elapsed) => {
            warn!("EndSession timed out; skipping afterLease wait");
            return;
        }
    }

    // afterLease can be slow (hooks have their own timeouts); bound the overall wait
    // but let the exporter drive completion. Done when it returns to Available/Offline
    // or a hook reaches a terminal failed state. An afterLease `on_failure: warn`
    // warning is delivered on the log stream (the status message is overwritten by the
    // trailing `request_release` Available report, so a poll here can miss it) — the
    // `spawn_controller` log task renders it as `Warning: …`.
    // afterLease hooks can be slow; match the beforeLease budget so a legitimately slow cleanup
    // hook isn't cut off (the lease is released either way once this returns).
    let deadline = Instant::now() + Duration::from_secs(300);
    while Instant::now() < deadline {
        let status =
            match tokio::time::timeout(RPC_TIMEOUT, client.get_status(GetStatusRequest {})).await {
                Ok(Ok(resp)) => ExporterStatus::try_from(resp.into_inner().status)
                    .unwrap_or(ExporterStatus::Unspecified),
                // Session/tunnel gone or RPC timed out — afterLease is done (or the
                // exporter exited); stop waiting.
                _ => {
                    debug!("end_session: status stream gone; afterLease assumed complete");
                    break;
                }
            };
        trace!(?status, "end_session: polled exporter status");
        match status {
            // afterLease ran but the hook failed: previously indistinguishable from
            // a clean completion — surface it so a failed afterLease is diagnosable.
            ExporterStatus::AfterLeaseHookFailed | ExporterStatus::BeforeLeaseHookFailed => {
                warn!(?status, "afterLease hook reported a terminal failed status");
                break;
            }
            // Clean: the exporter returned to Available/Offline (afterLease done).
            ExporterStatus::Available | ExporterStatus::Offline => {
                debug!(?status, "afterLease hook completed; exporter idle");
                break;
            }
            _ => {}
        }
        tokio::time::sleep(Duration::from_millis(200)).await;
    }
    if Instant::now() >= deadline {
        warn!("afterLease wait timed out (300s); releasing lease anyway");
    }
}

/// Direct mode (`jmp shell --tls-grpc HOST:PORT`): connect a shell/command to a
/// standalone exporter with no controller/router/lease. Python does not proxy —
/// it hands `JUMPSTARTER_HOST=host:port` (+ `JMP_GRPC_INSECURE`/`JMP_GRPC_PASSPHRASE`)
/// to the child, whose `j`/driver client dials the exporter's `ExporterService`
/// directly (`shell.py:_shell_direct_async`, `DirectLease`). `drivers_allow` is the
/// literal `UNSAFE` (DirectLease uses `unsafe=True`).
pub async fn run_direct(
    address: &str,
    insecure: bool,
    passphrase: Option<&str>,
    exporter_logs: bool,
    command: &Option<Vec<String>>,
) -> Result<i32, ClientError> {
    let log_task = if exporter_logs {
        let task = crate::exporter_logs::spawn_direct(
            address.to_string(),
            insecure,
            passphrase.map(str::to_string),
        );
        // Let the stream connect and replay buffered (pre-connect) logs before the
        // command runs — mirrors Python's "stream logs, then run command" order.
        tokio::time::sleep(std::time::Duration::from_millis(500)).await;
        Some(task)
    } else {
        None
    };

    let result = spawn_child(
        command, address, address, "UNSAFE", insecure, passphrase, None,
    )
    .await;

    if let Some(task) = log_task {
        // Brief flush window for any trailing (afterLease) lines before aborting.
        tokio::time::sleep(std::time::Duration::from_millis(200)).await;
        task.abort();
    }
    result
}

/// The exporter-context env the shell exports for a remote lease (#53): `JMP_EXPORTER`,
/// `JMP_LEASE`, and `JMP_EXPORTER_LABELS` (a sorted, comma-joined `k=v` list). Code inside the shell
/// reads these via `jumpstarter.utils.env.env_with_metadata()` / `ExporterMetadata.from_env()`.
/// `None` in direct mode (no controller lease).
pub struct LeaseEnv {
    pub exporter: String,
    pub lease: String,
    pub labels: String,
}

/// Spawn the shell/command with the `JUMPSTARTER_HOST` env contract
/// (`common/utils.py:launch_shell`). `host` is the bare socket path (controller/
/// local mode) or `host:port` (direct mode); `context` labels the decorated prompt
/// (the exporter name); `passphrase` sets `JMP_GRPC_PASSPHRASE` for a standalone
/// exporter. `lease_env` exports the exporter-context vars for a remote lease (#53).
/// Reused by the CLI's direct and local-exporter shell paths.
pub async fn spawn_child(
    command: &Option<Vec<String>>,
    host: &str,
    context: &str,
    drivers_allow: &str,
    insecure: bool,
    passphrase: Option<&str>,
    lease_env: Option<&LeaseEnv>,
) -> Result<i32, ClientError> {
    let mut cmd = match command {
        Some(argv) if !argv.is_empty() => {
            let mut c = tokio::process::Command::new(&argv[0]);
            c.args(&argv[1..]);
            c
        }
        // An interactive shell gets a jumpstarter-decorated prompt (`{cwd} ⚡{context} ➤`).
        _ => decorated_shell(context),
    };

    cmd.env("JUMPSTARTER_HOST", host)
        .env("JMP_DRIVERS_ALLOW", drivers_allow)
        .env("_JMP_SUPPRESS_DRIVER_WARNINGS", "1");
    if insecure {
        cmd.env("JMP_GRPC_INSECURE", "1");
    }
    if let Some(p) = passphrase.filter(|p| !p.is_empty()) {
        cmd.env(jumpstarter_config::env::JMP_GRPC_PASSPHRASE, p);
    }
    if let Some(le) = lease_env {
        cmd.env("JMP_EXPORTER", &le.exporter);
        if !le.lease.is_empty() {
            cmd.env("JMP_LEASE", &le.lease);
        }
        if !le.labels.is_empty() {
            cmd.env("JMP_EXPORTER_LABELS", &le.labels);
        }
    }

    let status = cmd
        .status()
        .await
        .map_err(|e| ClientError::Config(format!("failed to spawn shell/command: {e}")))?;

    // Signal-terminated child -> 2 (matches `j`'s cancellation exit code).
    Ok(status.code().unwrap_or(2))
}

/// Build the interactive `$SHELL` command with a jumpstarter-decorated prompt —
/// gray cwd, yellow `⚡`, white exporter name, yellow `➤` — suppressing rc/profile
/// files so the prompt actually takes effect (`common/utils.py:launch_shell`).
/// bash/zsh/fish each get their native prompt syntax; any other shell just inherits
/// the env contract with its default prompt.
fn decorated_shell(context: &str) -> tokio::process::Command {
    let shell = std::env::var("SHELL").unwrap_or_else(|_| "/bin/sh".to_string());
    let shell_name = std::path::Path::new(&shell)
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or("sh")
        .to_string();
    let mut c = tokio::process::Command::new(&shell);
    if shell_name.ends_with("bash") {
        // `\[..\]` wraps non-printing escapes; `\W` is the basename of $PWD.
        c.arg("--norc").arg("--noprofile").env(
            "PS1",
            format!("\\[\\e[90m\\]\\W \\[\\e[93m\\]⚡\\[\\e[97m\\]{context} \\[\\e[93m\\]➤\\[\\e[0m\\] "),
        );
    } else if shell_name == "zsh" {
        c.arg("--no-rcs")
            .args(["-o", "inc_append_history", "-o", "share_history"])
            .env(
                "PS1",
                format!("%F{{8}}%1~ %F{{yellow}}⚡%F{{white}}{context} %F{{yellow}}➤%f "),
            );
        if std::env::var_os("HISTFILE").is_none() {
            if let Some(home) = std::env::var_os("HOME") {
                c.env("HISTFILE", std::path::Path::new(&home).join(".zsh_history"));
            }
        }
    } else if shell_name == "fish" {
        c.arg("--init-command").arg(format!(
            "function fish_prompt; set_color grey; printf \"%s\" (basename $PWD); \
             set_color yellow; printf \"⚡\"; set_color white; printf \"{context}\"; \
             set_color yellow; printf \"➤ \"; set_color normal; end"
        ));
    }
    c
}

/// Log the outcome of a best-effort lease release. The controller auto-releases a lease when its
/// session ends, so an explicit release after the command often races and returns
/// `FailedPrecondition: ... already been released` — that is the expected no-op, logged at debug.
/// Any other failure is a real warning.
fn log_release_outcome(name: &str, err: &impl std::fmt::Display) {
    let msg = err.to_string();
    if msg.contains("already") && msg.contains("released") {
        debug!(lease = %name, "lease already released by the controller (expected no-op)");
    } else {
        warn!(lease = %name, error = %msg, "failed to release lease");
    }
}
